# 2. The Technician Assignment Model

**This is the reproducibility notebook.** It builds the model from its sets and
parameters through to a solved, verified dispatch plan — the run that produces
the paper's headline numbers.

`05_manuscript_results.ipynb` is a different thing: it reports finished values as
paste-ready tables. This notebook is where the model is actually constructed.

---

## What this model is

**Class:** identical parallel machine scheduling, $P \,\|\, \sum_j w_j C_j$, with
crew capacity, a shift-duration limit and a spatial-equity constraint.

Each fault is a job. Each crew is a machine. A crew dispatches from the Hohoe
base, repairs, and returns before its next job, so job $j$ occupies
$s_j = 2t_j + r_j$ of a crew's shift.

This is an **assignment** problem with a service order. There is no inter-site
travel, no travel matrix, no subtour elimination and no routing. The order
matters only because a customer later in a crew's shift waits longer — that is
scheduling, not routing.

## What changed from the original model, and why it had to

The published model minimised **total travel time**. Under the coverage
constraint that every fault is served exactly once, that objective is a
**constant** — proved in §2.3 below, on this data. Every feasible assignment
scored identically, so the solver was not choosing a good plan; it was returning
the first feasible point it found.

The objective is now **response time**: hours from shift start until a crew
reaches each customer. That depends on which crew takes the fault and on how many
jobs precede it, so no such collapse exists.

In [1]:
import pathlib, sys
ROOT = pathlib.Path.cwd().parent if pathlib.Path.cwd().name == "notebooks" else pathlib.Path.cwd()
sys.path.insert(0, str(ROOT))

import math
import time

import pandas as pd
import pulp

from src.config import Params
from src.data import (job_times, load_faults, priority_weights, zone_floors)
from src.model import build_and_solve
from src.solver import Column, enumerate_columns, solve
from src.verify import VerificationError, check_solution, objective_varies

print(f"PuLP {pulp.__version__} | solver: CBC (bundled)")

PuLP 3.3.2 | solver: CBC (bundled)


## 2.1 Sets and parameters

In [2]:
P = Params()
faults = load_faults()

J = list(range(len(faults)))                      # faults
I = list(range(P.n_crews))                        # crews
K = list(range(P.capacity))                       # service positions in a shift

travel = faults["Travel_time_hours"].tolist()     # t_j, base -> fault, one way
repair = faults["Repair_time_hours"].tolist()     # r_j, on site
zones = faults["Zone"].tolist()                   # z_j
ids = faults["Fault_ID"].tolist()

print(f"J  faults    : {len(J)}")
print(f"I  crews     : {len(I)}  (3 technicians each, moving as one unit)")
print(f"K  positions : {len(K)}  (max faults per crew per shift, Q)")
print(f"H  shift     : {P.shift_hours} h")
print(f"theta        : {P.theta}   (equity tolerance, defined on [1, inf))")
print(f"alpha, beta  : {P.alpha}, {P.beta}   (efficiency / equity weights)")

J  faults    : 15
I  crews     : 5  (3 technicians each, moving as one unit)
K  positions : 3  (max faults per crew per shift, Q)
H  shift     : 8.0 h
theta        : 1.6   (equity tolerance, defined on [1, inf))
alpha, beta  : 0.7, 0.3   (efficiency / equity weights)


In [3]:
# Job occupancy. The crew drives out, repairs, and drives back to base before
# its next job, so s_j = 2*t_j + r_j.
#
# The RETURN leg occupies the crew but no customer waits on it, so it enters the
# shift constraint and NEVER the objective. Getting that wrong would charge
# customers for the crew's drive home.
s = job_times(faults, round_trip=P.round_trip)

print("| Fault | Town | Zone | t_j (h) | r_j (h) | s_j = 2t+r |")
print("|---|---|---|---|---|---|")
for j in J:
    print(f"| {ids[j]} | {faults['Town'][j]} | {zones[j]} | {travel[j]:.3f} "
          f"| {repair[j]:.2f} | {s[j]:.3f} |")
print(f"\ntotal crew-time required : {sum(s):.3f} h")
print(f"total crew-time available: {P.n_crews * P.shift_hours:.1f} h")
print(f"utilisation              : {100 * sum(s) / (P.n_crews * P.shift_hours):.1f}%")

| Fault | Town | Zone | t_j (h) | r_j (h) | s_j = 2t+r |
|---|---|---|---|---|---|
| F1 | Afadzo South | Far | 1.017 | 1.00 | 3.034 |
| F2 | Liati | Near | 0.550 | 0.75 | 1.850 |
| F3 | Golokwati | Far | 0.500 | 0.75 | 1.750 |
| F4 | Agome yo | Near | 0.517 | 2.00 | 3.034 |
| F5 | Logba | Far | 1.000 | 2.00 | 4.000 |
| F6 | Afadzo South | Far | 1.017 | 0.33 | 2.364 |
| F7 | Akpafu | Near | 0.417 | 0.33 | 1.164 |
| F8 | Agome yo | Near | 0.517 | 0.33 | 1.364 |
| F9 | Santrokofi | Near | 0.333 | 2.00 | 2.666 |
| F10 | Wli | Far | 0.483 | 4.00 | 4.966 |
| F11 | Ve-Dator | Near | 0.600 | 0.75 | 1.950 |
| F12 | Ve -Gbodome | Far | 0.667 | 4.00 | 5.334 |
| F13 | Ve- Kobenu | Near | 0.583 | 0.33 | 1.496 |
| F14 | Agome yo | Near | 0.517 | 0.75 | 1.784 |
| F15 | Gbi - Kledzo | Near | 0.267 | 0.75 | 1.284 |

total crew-time required : 38.040 h
total crew-time available: 40.0 h
utilisation              : 95.1%


> Two features of this instance drive everything downstream.
>
> **Capacity exactly equals demand** — 5 crews × 3 faults = 15 faults, so every
> crew must take exactly three.
>
> **Utilisation is 95.1%.** Together these make the day a tight bin-packing
> problem, which is why ad-hoc dispatch usually produces a plan that does not fit
> the shift at all (notebook 04).

In [4]:
# Zone floors: each zone's mean ONE-WAY travel -- the geographic floor a crew
# cannot beat. Equity is measured as wait ABOVE this, because Far towns are
# further by construction and equalising raw means could only be achieved by
# delaying Near-zone customers.
floors = zone_floors(faults)
weights = priority_weights(faults, P.high_priority_weight)   # 1.0: confound control

for z, v in sorted(floors.items()):
    members = [j for j in J if zones[j] == z]
    print(f"{z:>5}: {len(members)} faults, mean one-way travel {v:.4f} h")
print(f"\npriority weights all {P.high_priority_weight} -- four of the five High-priority")
print("faults are in Far towns, so any higher weight would move the Far/Near")
print("ratio with no equity mechanism doing the work.")

  Far: 6 faults, mean one-way travel 0.7807 h
 Near: 9 faults, mean one-way travel 0.4779 h

priority weights all 1.0 -- four of the five High-priority
faults are in Far towns, so any higher weight would move the Far/Near
ratio with no equity mechanism doing the work.


## 2.2 Decision variables

$y_{ijk} = 1$ if crew $i$ services fault $j$ at position $k$ of its shift.

$R_j \ge 0$ is the **response time** of fault $j$: hours from shift start until a
crew arrives. $G \ge 0$ is the inter-zone equity gap, $C_{\max}$ the makespan.

In [5]:
print(f"y_ijk : {len(I)} x {len(J)} x {len(K)} = {len(I)*len(J)*len(K)} binaries")
print(f"R_j   : {len(J)} continuous (response time)")
print("G, Cmax, Rmax : 3 continuous")
print("\nThe full formulation is in src/model.py. It is the statement a referee")
print("expects to see; section 2.5 explains why a different ENCODING is used to")
print("solve it, and section 2.7 checks the two agree.")

y_ijk : 5 x 15 x 3 = 225 binaries
R_j   : 15 continuous (response time)
G, Cmax, Rmax : 3 continuous

The full formulation is in src/model.py. It is the statement a referee
expects to see; section 2.5 explains why a different ENCODING is used to
solve it, and section 2.7 checks the two agree.


## 2.3 Why the original objective could not work

Under coverage, $\sum_i\sum_j t_j x_{ij} = \sum_j t_j\left(\sum_i x_{ij}\right)
= \sum_j t_j$, because $t_j$ depends only on *which fault*, never on *which crew*
or *in what order*.

This holds whether travel is counted one-way or as a round trip, so it is a
property of the arithmetic rather than of the dispatch protocol. The cell below
demonstrates it, and the one after shows what the response-time objective does
on the same plans.

In [6]:
cols = enumerate_columns(travel, s, zones, weights, P.capacity, P.shift_hours)
total_possible = sum(math.perm(len(J), k) for k in range(1, P.capacity + 1))
print(f"crew workloads: {len(cols)} shift-feasible of {total_possible} possible\n")

eff = Params(alpha=1.0, beta=0.0)
lo = solve(travel, repair, zones, floors, weights, eff, sense="min", columns=cols)
hi = solve(travel, repair, zones, floors, weights, eff, sense="max", columns=cols)
lo_mean = lo.reported["weighted_mean_response"]
hi_mean = hi.reported["weighted_mean_response"]
_, spread = objective_varies([lo_mean, hi_mean])

print("| Objective | Best feasible plan | Worst feasible plan | Spread |")
print("|---|---|---|---|")
print(f"| Total travel time (original) | {sum(travel):.3f} h | {sum(travel):.3f} h | **0.0%** |")
print(f"| Mean response time (this model) | {lo_mean:.4f} h | {hi_mean:.4f} h | **{spread:.1f}%** |")
print("\nA 0.0% spread means the objective cannot tell a good plan from a bad one.")

crew workloads: 1793 shift-feasible of 2955 possible



| Objective | Best feasible plan | Worst feasible plan | Spread |
|---|---|---|---|
| Total travel time (original) | 8.985 h | 8.985 h | **0.0%** |
| Mean response time (this model) | 2.3002 h | 3.9075 h | **69.9%** |

A 0.0% spread means the objective cannot tell a good plan from a bad one.


In [7]:
# The clearest witness: one crew's workload, worked in two orders.
def arrivals(seq):
    cum, out = 0.0, []
    for j in seq:
        out.append(cum + travel[j])
        cum += s[j]
    return out, cum

demo = [14, 7, 11]                      # F15, F8, F12
fwd, dur = arrivals(demo)
rev, dur_r = arrivals(demo[::-1])
print("| Service order | Arrival times (h) | Total customer waiting | Crew time |")
print("|---|---|---|---|")
print(f"| {' -> '.join(ids[j] for j in demo)} | {', '.join(f'{a:.3f}' for a in fwd)} "
      f"| **{sum(fwd):.3f} h** | {dur:.3f} h |")
print(f"| {' -> '.join(ids[j] for j in demo[::-1])} | {', '.join(f'{a:.3f}' for a in rev)} "
      f"| **{sum(rev):.3f} h** | {dur_r:.3f} h |")
print(f"\nIdentical crew time ({dur:.3f} h). Customer waiting differs by "
      f"{max(sum(fwd), sum(rev)) / min(sum(fwd), sum(rev)):.2f}x.")
print(f"Both score {sum(travel[j] for j in demo):.3f} h under the original objective.")

| Service order | Arrival times (h) | Total customer waiting | Crew time |
|---|---|---|---|
| F15 -> F8 -> F12 | 0.267, 1.801, 3.315 | **5.383 h** | 7.982 h |
| F12 -> F8 -> F15 | 0.667, 5.851, 6.965 | **13.483 h** | 7.982 h |

Identical crew time (7.982 h). Customer waiting differs by 2.50x.
Both score 1.451 h under the original objective.


## 2.4 Constraints

| | Constraint | Meaning |
|---|---|---|
| C1 | $\sum_i\sum_k y_{ijk} = 1\ \forall j$ | every fault served exactly once |
| C2 | $\sum_j y_{ijk} \le 1\ \forall i,k$ | one fault per position slot |
| C3 | $\sum_j y_{ijk} \ge \sum_j y_{ij,k+1}$ | no idle gaps in a crew's shift |
| C4 | $\sum_j\sum_k s_j y_{ijk} \le H\ \forall i$ | shift-duration limit |
| C5 | $R_j$ pinned by **equality** to its arrival time | see below |
| C6 | $G \ge \lvert E^F - E^N\rvert$ | two-sided equity gap on **excess wait** |
| C7 | $E^F \le \theta E^N$ and $E^N \le \theta E^F$ | two-sided equity cap, $\theta \ge 1$ |

**C5 is pinned by equality, not minimality.** With only a lower bound, $R_j$ is
driven to the true arrival time *solely* by the efficiency term; whenever the
objective stops pushing it down (β = 1, or an ε-constraint) the solver inflates
$R_j$ and buys equity with fiction. An earlier version of this project had
exactly that defect and reported a gap of 0.000 against a true 0.588 h.

**C6 is two-sided** for the same reason: a one-sided constraint lets the solver
manufacture parity by *delaying Near-zone* customers.

**C7 requires θ ≥ 1.** Two-sided, $E^F \le \theta E^N$ and $E^N \le \theta E^F$
imply $E^F \le \theta^2 E^F$, so θ < 1 is identically infeasible. `src/config.py`
rejects it rather than letting the model report a misleading `Infeasible`.

## 2.5 Solution method: set partitioning over enumerated crew workloads

The constraints above describe the model. Solving it as a compact MILP with
big-M arrival constraints works but is slow — minutes at n = 15.

Because $Q = 3$, the complete set of ordered crew workloads is
$15 + 15\cdot14 + 15\cdot14\cdot13 = 2{,}955$, small enough to enumerate
exhaustively and price **exactly** in closed form. Three consequences:

1. The column set is **complete**, so the set-partitioning optimum is the exact
   optimum — not a heuristic, not a relaxation.
2. There is **no big-M**, so no weak relaxation and no slow solve.
3. Response times are **computed**, never chosen by the solver, which makes the
   decoupled-equity defect described above *structurally impossible*.

Point 3 is the important one and belongs in the paper.

In [8]:
t0 = time.time()
cols = enumerate_columns(travel, s, zones, weights, P.capacity, P.shift_hours)
enum_s = time.time() - t0
print(f"enumerated {len(cols)} shift-feasible workloads of {total_possible} "
      f"possible in {enum_s:.3f} s")
print(f"  ({100 * len(cols) / total_possible:.1f}% of orderings fit an {P.shift_hours:g} h shift)\n")

c = cols[0]
print("each column carries its exact cost, computed in plain Python:")
print(f"  faults   {[ids[j] for j in c.faults]}")
print(f"  arrivals {[round(a, 3) for a in c.arrivals]}")
print(f"  duration {c.duration:.3f} h")
print(f"  cost     {c.cost:.4f}  (sum of w_j * arrival_j)")

enumerated 1793 shift-feasible workloads of 2955 possible in 0.007 s
  (60.7% of orderings fit an 8 h shift)

each column carries its exact cost, computed in plain Python:
  faults   ['F1']
  arrivals [1.017]
  duration 3.034 h
  cost     1.0170  (sum of w_j * arrival_j)


## 2.6 Solve

In [9]:
sol = solve(travel, repair, zones, floors, weights, P, columns=cols)
meta = sol.columns

print(f"status          : {sol.status}")
print(f"search completed: {meta['search_completed']}   "
      f"(read from CBC's log, not inferred from PuLP's status)")
print(f"objective       : {sol.objective:.6f}")
print(f"solve time      : {meta['solve_seconds']:.3f} s")
print(f"model size      : {sol.n_variables} variables, {sol.n_constraints} constraints")

status          : Optimal
search completed: True   (read from CBC's log, not inferred from PuLP's status)
objective       : 1.658273
solve time      : 0.331 s
model size      : 1794 variables, 20 constraints


## 2.7 Verification — the step that must never be skipped

Both defects that nearly sank this project shared one shape: a quantity that
*looked* optimised but was decoupled from the decision. Neither was caught by
reading the model. Both are caught instantly by taking the schedule the solver
chose, recomputing every reported number from it in plain Python, and asserting
agreement.

If the cell below raises, **no number in this notebook can be used**.

In [10]:
rc = check_solution(
    sol.schedule, sol.reported,
    travel=travel, job_time=s, zones=zones, floors=floors, weights=weights,
    n_faults=len(J), status=sol.status,
    shift_hours=P.shift_hours, capacity=P.capacity)
print("VERIFIED at 1e-6: coverage, solver status, shift and capacity limits, and")
print("every reported quantity recomputed independently from the schedule.")

VERIFIED at 1e-6: coverage, solver status, shift and capacity limits, and
every reported quantity recomputed independently from the schedule.


## 2.8 The dispatch plan

In [11]:
print("| Crew | Faults in service order | Towns | Arrival times (h) | Crew time (h) |")
print("|---|---|---|---|---|")
for crew, seq in sorted(sol.schedule.items()):
    print(f"| {crew + 1} | {' -> '.join(ids[j] for j in seq)} "
          f"| {', '.join(faults['Town'][j] for j in seq)} "
          f"| {', '.join(f'{rc.response[j]:.3f}' for j in seq)} "
          f"| {rc.loads[crew]:.3f} |")

obj = P.alpha * rc.weighted_mean_response + P.beta * rc.equity_gap
print()
print("| Metric | Value |")
print("|---|---|")
for k, v in [("Mean response time", f"{rc.weighted_mean_response:.4f} h"),
             ("Worst customer wait", f"{rc.max_response:.4f} h"),
             ("Makespan (longest crew day)", f"{rc.makespan:.4f} h"),
             ("Mean response, Near", f"{rc.zone_mean['Near']:.4f} h"),
             ("Mean response, Far", f"{rc.zone_mean['Far']:.4f} h"),
             ("Excess wait, Near", f"{rc.zone_excess['Near']:.4f} h"),
             ("Excess wait, Far", f"{rc.zone_excess['Far']:.4f} h"),
             ("Equity gap", f"{rc.equity_gap:.4f} h"),
             (f"Objective (alpha={P.alpha}, beta={P.beta})", f"{obj:.4f}")]:
    print(f"| {k} | {v} |")

| Crew | Faults in service order | Towns | Arrival times (h) | Crew time (h) |
|---|---|---|---|---|
| 1 | F3 -> F1 -> F9 | Golokwati, Afadzo South, Santrokofi | 0.500, 2.767, 5.117 | 7.450 |
| 2 | F6 -> F11 -> F4 | Afadzo South, Ve-Dator, Agome yo | 1.017, 2.964, 4.831 | 7.348 |
| 3 | F7 -> F2 -> F10 | Akpafu, Liati, Wli | 0.417, 1.714, 3.497 | 7.980 |
| 4 | F13 -> F14 -> F5 | Ve- Kobenu, Agome yo, Logba | 0.583, 2.013, 4.280 | 7.280 |
| 5 | F15 -> F8 -> F12 | Gbi - Kledzo, Agome yo, Ve -Gbodome | 0.267, 1.801, 3.315 | 7.982 |

| Metric | Value |
|---|---|
| Mean response time | 2.3389 h |
| Worst customer wait | 5.1170 h |
| Makespan (longest crew day) | 7.9820 h |
| Mean response, Near | 2.1897 h |
| Mean response, Far | 2.5627 h |
| Excess wait, Near | 1.7118 h |
| Excess wait, Far | 1.7820 h |
| Equity gap | 0.0702 h |
| Objective (alpha=0.7, beta=0.3) | 1.6583 |


## 2.9 Cross-check: does the compact MILP agree?

The set-partitioning encoding is only legitimate if it solves the *same model*.
The compact big-M formulation in `src/model.py` is checked against it here.

It is run at **reduced size**, because the big-M encoding takes minutes at
n = 15 while set partitioning takes a fraction of a second — which is itself the
argument for the encoding, and worth reporting in the paper.

In [12]:
N_SMALL = 7
sub = faults.head(N_SMALL)
t_s = sub["Travel_time_hours"].tolist()
r_s = sub["Repair_time_hours"].tolist()
z_s = sub["Zone"].tolist()
f_s = zone_floors(sub)
w_s = priority_weights(sub, P.high_priority_weight)
P_s = Params(n_crews=3)

sp = solve(t_s, r_s, z_s, f_s, w_s, P_s)
t0 = time.time()
mi = build_and_solve(t_s, r_s, z_s, f_s, w_s, P_s, time_limit=600)
milp_s = time.time() - t0

print(f"n = {N_SMALL}, m = {P_s.n_crews}\n")
print(f"  set partitioning : {sp.status:<10} objective {sp.objective:.6f}"
      f"  in {sp.columns['solve_seconds']:.3f} s")
print(f"  compact MILP     : {mi.status:<10} objective {mi.objective:.6f}"
      f"  in {milp_s:.3f} s")
gap = abs(sp.objective - mi.objective)
print(f"\n  difference: {gap:.2e}")
assert gap < 1e-6, "the two encodings disagree -- stop and investigate"
print("  AGREE to 1e-6: the two encodings describe the same model.")

n = 7, m = 3

  set partitioning : Optimal    objective 1.335500  in 0.036 s
  compact MILP     : Optimal    objective 1.335500  in 2.042 s

  difference: 2.22e-16
  AGREE to 1e-6: the two encodings describe the same model.


## 2.10 Save the plan

In [13]:
rows = []
for crew, seq in sorted(sol.schedule.items()):
    for pos, j in enumerate(seq, start=1):
        rows.append({"Crew": crew + 1, "Position": pos, "Fault_ID": ids[j],
                     "Town": faults["Town"][j], "Zone": zones[j],
                     "Fault_Type": faults["Fault_type"][j],
                     "Priority": faults["Priority"][j],
                     "Travel_hrs": travel[j], "Repair_hrs": repair[j],
                     "Job_occupancy_hrs": round(s[j], 3),
                     "Response_hrs": round(rc.response[j], 3)})
out = pd.DataFrame(rows)
out.to_csv(ROOT / "result" / "optimization_results.csv", index=False)
print(f"written: result/optimization_results.csv ({len(out)} rows)")
out

written: result/optimization_results.csv (15 rows)


,Crew,Position,Fault_ID,Town,Zone,Fault_Type,Priority,Travel_hrs,Repair_hrs,Job_occupancy_hrs,Response_hrs
0,1,1,F3,Golokwati,Far,Transformer maintenance,High,0.500,0.75,1.750,0.500
1,1,2,F1,Afadzo South,Far,Pole replacement,High,1.017,1.00,3.034,2.767
2,1,3,F9,Santrokofi,Near,Network line extension,Normal,0.333,2.00,2.666,5.117
3,2,1,F6,Afadzo South,Far,Cable joining and termination,High,1.017,0.33,2.364,1.017
4,2,2,F11,Ve-Dator,Near,Transformer maintenance,Normal,0.600,0.75,1.950,2.964
5,2,3,F4,Agome yo,Near,Network line extension,Normal,0.517,2.00,3.034,4.831
6,3,1,F7,Akpafu,Near,Cable joining and termination,Normal,0.417,0.33,1.164,0.417
7,3,2,F2,Liati,Near,Transformer maintenance,Normal,0.550,0.75,1.850,1.714
8,3,3,F10,Wli,Far,Transformer installation,High,0.483,4.00,4.966,3.497
9,4,1,F13,Ve- Kobenu,Near,Cable joining and termination,Normal,0.583,0.33,1.496,0.583
